src/Shrubbery/Generic.hs

```diff
@@ -0,0 +1,221 @@
+-}
+genericFromUnion :: (Generic a, GenericUnion (Rep a))
+                 => Union (GenericBranchTypes (Rep a))
+                 -> a
+genericFromUnion (Union index t) =
+  genericUnifyWithIndex index t
+
+{- |
+  'GenericBranchTypes' should be as the implementation of the 'BranchTypes' family
+  for a type when either 'genericDissect' or 'genericUnifyWithIndex' are used.
+
+  For example:
+
+  @
+    type instance BranchTypes MyType = GenericBranchTypes (Rep MyType)
+  @
+-}
+type family GenericBranchTypes (rep :: Type -> Type) :: [Type] where
+  GenericBranchTypes (M1 i c f) = GenericBranchTypes f
+  GenericBranchTypes (a :+: b) = AppendTypes (GenericBranchTypes a) (GenericBranchTypes b)
+  GenericBranchTypes (K1 i c) = '[c]
+  GenericBranchTypes U1 = '[()]
+
+{- |
+  'GenericUnion' provides sum-type analysis for GHC Generics representation of
+  sum types. By deriving 'Generic' for your traditional Haskell sum types you
+  can the use 'genericUnifyWithIndex', 'genericDissect' and
+  'GenericBranchTypes' in your implementations of 'Unification', 'Dissection'
+  and 'BranchTypes.
+
+  Warning: See the discussion in the module documentation for a reason you
+  might not actually want to do this.
+-}
+class GenericUnion (rep :: Type -> Type) where
+  {- |
+    'genericToUnion' allows a sum type with a 'Generic' representation of
+    multiple single-argument constructors to be converted to a 'Union' of its
+    'GenericBranchTypes'.
+
+    Warning: See the discussion in the module documentation for a reason you
+    might not actually want to do this.
+  -}
+  genericToUnion :: rep p -> Union (GenericBranchTypes rep)
+  constructorForIndex :: BranchIndex t (GenericBranchTypes rep) -> t -> rep p
+
+instance GenericUnion f => GenericUnion (M1 i c f) where
+  genericToUnion = genericToUnion . unM1
+  constructorForIndex index = M1 . constructorForIndex index
+
+instance GenericUnion (K1 i c) where
+  genericToUnion k1 =
+    -- 'firstIndexOfType' is always 0 here because there is only one type in
+    -- the list
+    Union firstIndexOfType (unK1 k1)
+
+  {-
+    The type in this case is
+    @BranchIndex t (GenericBranchTypes (K1 i c)) -> t -> K1 i c p@
+
+    Which expands to
+    @BranchIndex t '[c] -> t -> K1 i c p@
+
+    Given that a @BranchIndex@ can only be constructed when the type it
+    picks out is actually in the list of types being indexes, we can conclude
+    that @t ~ c@, even though GHC cannot conclude this.
+
+    This and the fact that the type of @K1@ is @c -> K1 i c p@ is enough to
+    allow us to conclude @K1@ has the required type and @unsafeCoerce@ is, in
+    fact, safe here.
+  -}
+  constructorForIndex _ = unsafeCoerce K1
+
+instance GenericUnion U1 where
+  genericToUnion _ =
+    -- 'firstIndexOfType' is always 0 here because there is only one type in
+    -- the list
+    Union firstIndexOfType ()
+
+  {-
+    Although there is no need to use @unsafeCoerce@ here, the same reasoning
```

Comment:
Do we need the note about unsafeCoerce here? I got twisted around looking for it, thinking that it wasn't strictly needed but the comment was around consistency with the K1 instance.